src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -556,14 +556,20 @@ object desugar {
       DefDef(nme.ANON_FUN, Nil, params :: Nil, tpt, body).withMods(synthetic),
       Closure(Nil, Ident(nme.ANON_FUN), EmptyTree))
 
-  /** Expand partial function
+  /** If `nparams` == 1, expand partial function
+   *
    *       { cases }
    *  ==>
-   *       x$0 => x$0 match { cases }
+   *       x$1 => x$1 match { cases }
+   *
+   *  If `nparams` > 1, expand instead to
```

Comment:
scalac also allows pattern matching anonymous functions with an expected type of `Function0`:

```
% cat sandbox/test.scala
class Test {
  ({case _ if "".isEmpty => 0} : (() => Any)).apply()
}

%  qscalac -Xprint:typer sandbox/test.scala
[[syntax trees at end of                     typer]] // test.scala
package <empty> {
  class Test extends scala.AnyRef {
    def <init>(): Test = {
      Test.super.<init>();
      ()
    };
    ((() => () match {
  case _ if "".isEmpty() => 0
}): () => Any).apply()
  }
}
```